src/vote_graph.rs

```diff
@@ -0,0 +1,515 @@
+/// and vote data which is accumulated along edges.
+pub struct VoteGraph<H: Hash + Eq, V> {
+	entries: HashMap<H, Entry<H, V>>,
+	heads: HashSet<H>,
+	base: H,
+}
+
+impl<H, V> VoteGraph<H, V> where
+	H: Hash + Eq + Clone + Ord,
+	V: ::std::ops::AddAssign + Default + Clone,
+{
+	fn new(base_hash: H, base_number: usize) -> Self {
+		let mut entries = HashMap::new();
+		entries.insert(base_hash.clone(), Entry {
+			number: base_number,
+			ancestors: Vec::new(),
+			descendents: Vec::new(),
+			cumulative_vote: V::default(),
+		});
+
+		let mut heads = HashSet::new();
+		heads.insert(base_hash.clone());
+
+		VoteGraph {
+			entries,
+			heads,
+			base: base_hash,
+		}
+	}
+
+	/// Insert a vote with given value into the graph at given hash and number.
+	pub fn insert<C: Chain<H>>(&mut self, hash: H, number: usize, vote: V, chain: &C) -> Result<(), Error> {
+		match self.find_containing_nodes(hash.clone(), number) {
+			Some(containing) => if containing.is_empty() {
+				self.append(hash.clone(), number, chain)?;
+			} else {
+				self.introduce_branch(containing, hash.clone(), number);
+			},
+			None => {}, // this entry already exists
+		}
+
+		// update cumulative vote data.
+		// NOTE: below this point, there always exists a node with the given hash and number.
+		let mut inspecting_hash = hash;
+		loop {
+			let active_entry = self.entries.get_mut(&inspecting_hash)
+				.expect("vote-node and its ancestry always exist after initial phase; qed");
+
+			active_entry.cumulative_vote += vote.clone();
+
+			match active_entry.ancestor_node() {
+				Some(parent) => { inspecting_hash = parent },
+				None => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Find the best GHOST descendent of the given block.
+	/// Pass a closure used to evaluate the cumulative vote value.
+	///
+	/// The GHOST (hash, number) returned will be the block with highest number for which the
+	/// cumulative votes of descendents and itself causes the closure to evaluate to true.
+	///
+	/// This assumes that the evaluation closure is one which returns true for at most a single
+	/// descendent of a block, in that only one fork of a block can be "heavy"
+	/// enough to trigger the threshold.
+	pub fn find_ghost<'a, F>(&'a self, current_best: Option<(H, usize)>, condition: F) -> Option<(H, usize)>
+		where F: Fn(&V) -> bool
+	{
+		let entries = &self.entries;
+		let get_node = |hash: &_| -> &'a _ {
+			entries.get(hash)
+				.expect("node either base or referenced by other in graph; qed")
+		};
+
+		let mut node_key = current_best
+			.and_then(|(hash, number)| match self.find_containing_nodes(hash.clone(), number) {
+				None => Some(hash),
+				Some(ref x) if !x.is_empty() => Some(x[0].clone()),
+				Some(_) => None,
+			})
+			.and_then(|hash| get_node(&hash).ancestor_node())
+			.unwrap_or_else(|| self.base.clone());
+
+		let mut active_node = get_node(&node_key);
+
+		if !condition(&active_node.cumulative_vote) { return None }
```

Comment:
this function is defined to always return with a descendent of the supplied node. this question cuts to the crux of the issue mentioned in the PR description and comment on this function: that `find_ghost` isn't a proper GHOST but a weird "majority-GHOST" thing where you either have one descendent of a node that meets the condition or none. So we find the best block that meets the condition or none and we don't need to backtrack because of the majority condition.

Reply:
to note why the property of returning a descendent is useful: this function is needed for the vote-casting logic, where we want to cast votes either after some time has passed _or_ as soon as `GHOST(S) == GHOST(S + Y)` where `Y` could be any new valid votes to add to the vote set. 

And as the document shows, the GHOST block never backtracks as we include more votes. So by bookkeeping the current `GHOST(S)` node we can easily check `GHOST(S + Y)` with another call to the same function. Although it can still be optimized a bit it should still be pretty fast.